app/components/user/TwoFactor/RecoveryCodes/RecoveryCodeDialog.js

```diff
@@ -0,0 +1,204 @@
+// @flow
+
+import * as React from 'react';
+import { createFragmentContainer, graphql, commitMutation } from 'react-relay/compat';
+import { CopyToClipboard } from 'react-copy-to-clipboard';
+import GraphQLErrors from 'app/constants/GraphQLErrors';
+import Button from 'app/components/shared/Button';
+import Dialog from 'app/components/shared/Dialog';
+import RecoveryCodeList from 'app/components/RecoveryCodeList';
+import type { RecoveryCodeDialog_totp } from './__generated__/RecoveryCodeDialog_totp.graphql';
+import type { _RecoveryCodeDialogCreateMutationVariables, RecoveryCodeDialogCreateMutationResponse } from './__generated__/RecoveryCodeDialogCreateMutation.graphql';
+
+type RecoveryCodes = $PropertyType<RecoveryCodeDialog_totp, 'recoveryCodes'>;
+
+type Props = {
+  onRequestClose: Function,
+  totp: RecoveryCodeDialog_totp,
+  relay: Object
+};
+
+type State = {
+  generatingRecoveryCodes: boolean,
+  generatedRecoveryCodes: boolean,
+  recoveryCodes: RecoveryCodes
+};
+
+function recoveryCodeText(recoveryCodes: RecoveryCodes): ?string {
+  if (recoveryCodes && recoveryCodes.codes) {
+    return recoveryCodes.codes.reduce((memo, { code }) => memo.concat(code), []).join('\n');
+  }
+  return '';
+}
+
+class RecoveryCodesDialog extends React.PureComponent<Props, State> {
+  state = {
+    generatingRecoveryCodes: false,
+    generatedRecoveryCodes: false,
+    recoveryCodes: null
+  };
+
+  componentDidMount() {
+    // If we have a TOTP, but no recovery codes for it, immediately go generate some recovery codes.
+    if (this.props.totp && !this.props.totp.recoveryCodes) {
+      this.generateRecoveryCodes();
```

Comment:
@ticky We were having a look at this while talking about a refactor earlier today and we're not sure if it's still necessary? What do you think?

Reply:
That’s correct! This state should no longer be reachable. I was originally going to make the transition to recovery codes self-service but then Keith did some database magic instead :)